src/rust/ide/src/ide/integration.rs

```diff
@@ -1252,16 +1177,126 @@ impl Model {
+    /// visualization respective FRP endpoint.
+    fn visualization_update_handler
+    ( &self
+      , endpoint : frp::Any<(graph_editor::NodeId,visualization::Data)>
+      , node_id  : graph_editor::NodeId
+    ) -> impl FnMut(VisualizationUpdateData) -> futures::future::Ready<()> {
+        // TODO [mwu]
+        //  For now only JSON visualizations are supported, so we can just assume JSON data in the
+        //  binary package.
+        let logger = self.logger.clone_ref();
+        move |update| {
+            match Self::deserialize_visualization_data(update) {
+                Ok (data)  => endpoint.emit((node_id,data)),
+                Err(error) =>
+                // TODO [mwu]
+                //  We should consider having the visualization also accept error input.
+                    error!(logger, "Failed to deserialize visualization update. {error}"),
+            }
+            futures::future::ready(())
+        }
+    }
+
+    /// Create a controller-compatible description of the visualization based on the input received
+    /// from the graph editor endpoints.
+    fn prepare_visualization
+    (&self, node_id:graph_editor::NodeId, metadata:&Option<visualization::Metadata>)
+     -> FallibleResult<Visualization> {
+        use crate::model::module::QualifiedName;
+        let metadata = metadata.as_ref();
+
+        // TODO [mwu]
+        //   Currently it is not possible to:
+        //    * enter other module than the initial (namely, "Main")
+        //    * describe that visualization's expression wishes to be evaluated in any other
+        //      context.
+        //   Because of that for now we will just hardcode the `visualization_module` using
+        //   fixed defaults. In future this will be changed, then the editor will also get access
+        //   to the customised values.
+        let project_name:String  = self.project.name().into();
+        let module_name          = crate::ide::INITIAL_MODULE_NAME;
+        let visualisation_module = QualifiedName::from_segments(project_name,&[module_name])?;
+        let id                   = VisualizationId::new_v4();
+        let metadata_expression  = metadata.and_then(|m| m.preprocessor.as_ref().map(|e| e.to_string()));
+        let default_expression   = crate::constants::SERIALIZE_TO_JSON_EXPRESSION;
+        let expression           = metadata_expression.unwrap_or_else(||default_expression.into());
+        let ast_id               = self.get_controller_node_id(node_id)?;
+        Ok(Visualization{ast_id,expression,id,visualisation_module})
+    }
+
+    fn detach_visualization
+    ( &self
+    , node_id            : graph_editor::NodeId
+    , visualizations_map : SharedHashMap<graph_editor::NodeId,VisualizationId>
+    ) -> FallibleResult {
+        debug!(self.logger,"Node editor wants to detach visualization on {node_id}.");
+        let err            = || NoSuchVisualization(node_id);
```

Comment:
We have 4 lambdas here and they are using or not using space after pipes. This should be unified in the codebase. I'd go with no space. If you agree, please add it to styling docs too.

Reply:
Well, we put space everywhere, and it makes sense to me: consider `|foo,x,y|foo.blah(x,y)`, here separating arguments and body would be helpful